frameworks.qmd

```diff
@@ -1,119 +1,1785 @@
+graphs, optimization algorithms, hardware acceleration, and more that
+enable developers to quickly construct performant models. Understanding
+these frameworks is essential to leverage the power of deep learning
+across the spectrum from cloud to edge devices.
+
+ML frameworks handle much of the complexity of model development through
+high-level APIs and domain-specific languages that allow practitioners
+to quickly construct models by combining pre-made components and
+abstractions. For example, frameworks like TensorFlow and PyTorch
+provide Python APIs to define neural network architectures using layers,
+optimizers, datasets, and more. This enables rapid iteration compared to
+coding every model detail from scratch.
+
+A key capability offered by frameworks is distributed training engines
+that can scale model training across clusters of GPUs and TPUs. This
+makes it feasible to train state-of-the-art models with billions or
+trillions of parameters on vast datasets. Frameworks also integrate with
+specialized hardware like NVIDIA GPUs to further accelerate training via
+optimizations like parallelization and efficient matrix operations.
+
+In addition, frameworks simplify deploying finished models into
+production through tools like TensorFlow Serving for scalable model
+serving and TensorFlow Lite for optimization on mobile and edge devices.
+Other valuable capabilities include visualization, model optimization
+techniques like quantization and pruning, and monitoring metrics during
+training.
+
+Leading open source frameworks like TensorFlow, PyTorch, and MXNet power
+much of AI research and development today. Commercial offerings like
+Amazon SageMaker and Microsoft Azure Machine Learning integrate these
+open source frameworks with proprietary capabilities and enterprise
+tools.
+
+Machine learning engineers and practitioners leverage these robust
+frameworks to focus on high-value tasks like model architecture, feature
+engineering, and hyperparameter tuning instead of infrastructure. The
+goal is to efficiently build and deploy performant models that solve
+real-world problems.
+
+In this chapter, we will explore today\'s leading cloud frameworks and
+how they have adapted models and tools specifically for embedded and
+edge deployment. We will compare programming models, supported hardware,
+optimization capabilities, and more to fully understand how frameworks
+enable scalable machine learning from the cloud to the edge.
+
+## Framework Evolution
+
+Machine learning frameworks have evolved significantly over time to meet
+the diverse needs of machine learning practitioners and advancements in
+AI techniques. A few decades ago, building and training machine learning
+models required extensive low-level coding and infrastructure. Machine
+learning frameworks have evolved considerably over the past decade to
+meet the expanding needs of practitioners and rapid advances in deep
```

Comment:
This paragraph could be reorganized to be clearer and shorter. For example, these two sentences are very similar and could be condensed: "Machine learning frameworks have evolved significantly over time to meet the diverse needs of machine learning practitioners and advancements in AI techniques." and "Machine learning frameworks have evolved considerably over the past decade to meet the expanding needs of practitioners and rapid advances in deep learning techniques"